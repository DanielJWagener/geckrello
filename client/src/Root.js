// Simply creates a Redux store and passes it in to a provider tag.
// This logic was originally in index.js, but we've movied it here to
// make our testing process easier.
import React from "react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware, compose } from "redux";
import reduxThunk from "redux-thunk";
import reducers from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  reducers,
  {},
  composeEnhancers(applyMiddleware(reduxThunk))
);

export default ({ children }) => {
  return <Provider store={store}>{children}</Provider>;
};
