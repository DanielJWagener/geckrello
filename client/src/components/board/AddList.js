import React from "react";
import { connect } from "react-redux";
import { addList } from "../../actions";

class AddList extends React.Component {
  state = {
    mode: "prompt",
    listTitle: "",
    borderColor: "#555",
    spans: 0
  };

  // Inialize element height on grid. Its grid height is based on CSS grid spans, so we fetch that value from the DOM and set the number of spans accordingly
  addListRef = React.createRef();

  setSpans = () => {
    const height = this.addListRef.current.clientHeight;

    const spans = Math.floor(height / 10);

    this.setState({ spans: spans });
  };

  componentDidMount() {
    this.setSpans();
  }

  // Mode toggle. This component has two modes: "prompt" and "input." The "prompt" mode is the initial mode, basically a button with a label like "add new list." When the user clicks that button, the mode changes to "input," and the user can input and submit a new list and title.
  toggleMode = () => {
    if (this.state.mode === "prompt") {
      this.setState({ mode: "input" });
    } else {
      this.setState({ mode: "prompt" });
    }
  };

  // Submit handlers
  onFormSubmit = event => {
    event.preventDefault();

    // Validate input
    if (!this.state.listTitle) {
      alert("Please enter a title.");
      return false;
    }

    // Generate a unique key and listId in the event of duplicate list titles
    const generateListId = inputId => {
      if (!this.props.lists.find(list => list.listId === inputId)) {
        return inputId;
      } else {
        inputId = `${inputId}-duplicate`;
        return generateListId(inputId);
      }
    };
    let listId = generateListId(this.state.listTitle);

    // Submit input
    this.props.addList(this.state.listTitle, listId);

    // Clear input
    this.setState({ listTitle: "", inputValid: false, borderColor: "#555" });

    // Change mode back to "prompt"
    this.toggleMode();
  };

  // Controlled input and validation to ensure unique list titles
  handleChange = e => {
    this.setState({
      listTitle: e.target.value
    });

    e.target.value
      ? this.setState({ borderColor: "#4fa644" })
      : this.setState({ borderColor: "red" });
  };

  render() {
    if (this.state.mode === "prompt") {
      return (
        <div ref={this.addListRef} className="addList addList__prompt">
          <button onClick={this.toggleMode} className="addList__prompt-button">
            <span className="addList__prompt-plus">+</span>
            <span className="addList__prompt-label"> Add new list</span>
          </button>
        </div>
      );
    } else {
      return (
        <div
          className="addList addList__input"
          ref={this.addListRef}
          style={{ gridRowEnd: `span ${this.state.spans}` }}
        >
          <div className="addList__form">
            <form onSubmit={this.onFormSubmit} className="form">
              <div className="form__group">
                <input
                  type="text"
                  className="form form__input form__input--validation"
                  value={this.state.listTitle}
                  onChange={this.handleChange}
                  style={{ borderBottomColor: this.state.borderColor }}
                  placeholder="New list title..."
                  autoFocus
                />
              </div>
              <div className="form__group">
                <input
                  type="submit"
                  className="form__submit"
                  value="Add List"
                />
              </div>
            </form>
          </div>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  return { lists: state.lists };
};

export default connect(
  mapStateToProps,
  { addList }
)(AddList);
