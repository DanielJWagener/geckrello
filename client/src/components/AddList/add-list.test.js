import { shallow } from "enzyme";
import React from "react";
import AddList from "./add-list.component";
import Root from "../../Root";

it("renders AddList component", () => {
  expect(
    shallow(
      <Root>
        <AddList />
      </Root>
    )
  ).toMatchSnapshot();
});
