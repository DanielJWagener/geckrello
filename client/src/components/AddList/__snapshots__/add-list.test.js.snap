// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AddList component 1`] = `ShallowWrapper {}`;
